exist-core/src/main/java/org/exist/xquery/CountClause.java

```diff
@@ -0,0 +1,211 @@
+        while (prev != null) {
+            switch (prev.getType()) {
+                case LET, GROUPBY, FOR -> {
+                    return false;
+                }
+                case ORDERBY -> {
+                    return isDescending(((OrderByClause) prev).getOrderSpecs());
+                }
+            }
+            prev = prev.getPreviousClause();
+        }
+        return true;
+    }
+    private boolean isDescending(final OrderSpec[] orderSpecs) {
+        for (final OrderSpec orderSpec : orderSpecs) {
+            if ((orderSpec.getModifiers() & OrderSpec.DESCENDING_ORDER) == OrderSpec.DESCENDING_ORDER) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    @Override
+    public Sequence eval(final Sequence contextSequence, final Item contextItem) throws XPathException {
+        if (context.getProfiler().isEnabled()) {
+            context.getProfiler().start(this);
+            context.getProfiler().message(this, Profiler.DEPENDENCIES,
+                    "DEPENDENCIES", Dependency.getDependenciesName(this.getDependencies()));
+            if (contextSequence != null) {
+                context.getProfiler().message(this, Profiler.START_SEQUENCES,
+                        "CONTEXT SEQUENCE", contextSequence);
+            }
+            if (contextItem != null) {
+                context.getProfiler().message(this, Profiler.START_SEQUENCES,
+                        "CONTEXT ITEM", contextItem.toSequence());
+            }
+        }
+
+        context.expressionStart(this);
+
+        final Sequence resultSequence = new ValueSequence(unordered);
+
+        // update the count
+        count = count + step;
+
+        // Save the local variable stack
+        final LocalVariable mark = context.markLocalVariables(false);
+        try {
+
+            // Declare the count variable
+            final LocalVariable countVar = createVariable(varName);
+            countVar.setSequenceType(countVarType);
+            context.declareVariableBinding(countVar);
+
+            // set the binding for the count
+            countVar.setValue(new IntegerValue(count));
+
+            // eval the return expression on the window binding
+            resultSequence.addAll(returnExpr.eval(null));
+
+            // free resources
+            countVar.destroy(context, resultSequence);
+
+        } finally {
+            // restore the local variable stack
+            context.popLocalVariables(mark, resultSequence);
+        }
+
+        setActualReturnType(resultSequence.getItemType());
+
+        context.expressionEnd(this);
+        if (context.getProfiler().isEnabled()) {
+            context.getProfiler().end(this, "", resultSequence);
+        }
+
+        return resultSequence;
+    }
+
+    @Override
+    public Sequence postEval(Sequence seq) throws XPathException {
+        if (returnExpr instanceof FLWORClause) {
+            seq = ((FLWORClause)returnExpr).postEval(seq);
```

Comment:
Here the advanced `instanceof` could be used.

Reply:
Done.